sycl/source/detail/scheduler/commands.cpp

```diff
@@ -161,6 +226,190 @@ Command::Command(CommandType Type, QueueImplPtr Queue)
+      // event is registered by the runtime as a dependency The dependency on
+      // this occasion is an OCL event; so we build a virtual node in the graph
+      // with the event as the metadata for the node
+      std::string NodeName =
+          SH.nameWithAddressString("virtual_node", AddressStr);
+      // Node name is "virtual_node[<event_addr>]"
+      xpti::payload_t np(NodeName.c_str(), MAddress);
+      auto NodeEvent = xptiMakeEvent(NodeName.c_str(), &np,
+                                     (uint16_t)xpti_te::graph, xpti_at::active);
+      // Emit the virtual node first
+      xptiAddMetadata(NodeEvent, "kernel_name", NodeName.c_str());
+      xptiNotifySubscribers(MStreamID, (uint16_t)xpti_tp::node_create,
+                            cl::sycl::detail::g_sycl_graph_event, NodeEvent,
+                            nullptr);
+      // Create a new event for the edge
+      std::string EdgeName = SH.nameWithAddressString("Event", AddressStr);
+      xpti::payload_t p(EdgeName.c_str(), MAddress);
+      auto EdgeEvent = xptiMakeEvent(EdgeName.c_str(), &p,
+                                     (uint16_t)xpti_te::graph, xpti_at::active);
+      if (EdgeEvent && NodeEvent) {
+        // Source node represents the event and this event needs
+        // to be completed before target node can execute
+        xpti_td *TgtEvent = (xpti_td *)MTraceEvent;
+        EdgeEvent->source_id = NodeEvent->unique_id;
+        EdgeEvent->target_id = TgtEvent->unique_id;
+        xptiAddMetadata(EdgeEvent, "event", EdgeName.c_str());
+        xptiNotifySubscribers(MStreamID, (uint16_t)xpti_tp::edge_create,
+                              cl::sycl::detail::g_sycl_graph_event, EdgeEvent,
+                              nullptr);
+      }
+    }
+  }
+#endif
+}
+
+void Command::makeTraceEventProlog(void *MAddress) {
+#ifdef XPTI_ENABLE_INSTRUMENTATION
+  MTraceEventPrologComplete = true;
+  // Setup the member variables with information needed for event notification
+  MCommandNodeType = commandToNodeType(MType);
+  MCommandName = commandToName(MType);
+  xpti::utils::StringHelper SH;
+  MAddressString = SH.addressAsString(MAddress);
+  std::string CommandString =
+      SH.nameWithAddressString(MCommandName, MAddressString);
+
+  xpti::payload_t p(CommandString.c_str(), MAddress);
+  xpti_td *CmdTraceEvent = xptiMakeEvent(
+      CommandString.c_str(), &p, (uint16_t)xpti_te::graph, xpti_at::active);
+  if (CmdTraceEvent) {
+    MTraceEvent = (void *)CmdTraceEvent;
+    // If we are seeing this event again, then the instance ID will be greater
+    // than 1; in this case, we must skip sending a notification to create a
+    // node as this node has already been created. We return this value so the
+    // epilog method can be called selectively.
+    MFirstInstance = (CmdTraceEvent->instance_id == 1);
+  }
+#endif
+}
+
+void Command::makeTraceEventEpilog() {
+#ifdef XPTI_ENABLE_INSTRUMENTATION
+  if (MTraceEventPrologComplete) {
```

Comment:
How can it happen that we are here, but MTraceEventPrologComplete == false?
If it should not happen, please replace with assert.

Reply:
Done!